### PR TITLE
[sc58209] Add personal API keys APIs

### PR DESCRIPTION
**Pull request by qcastel**
This PR goes in pair with https://github.com/dataiku/dip/pull/15723

You will need to checkout this PR to run the integration tests of the other PR

### CODE REVIEW

**Review by FChataigner — APPROVED** [review]

File: dataikuapi/dss/admin.py [hunk 1]
```diff
@@ -1062,6 +1062,38 @@ def set_definition(self, definition):
...
+        self.id_ = id_
+
+    ########################################################
+    # Key description
+    ########################################################
+
+    def get_definition(self):
+        """
+        Get the API key's definition
+        :returns: the personal API key definition, as a JSON object
```
Comment (mhoarau) [inline dataikuapi/dss/admin.py:1081]: Missing a new line.

**Review by mhoarau — CHANGES_REQUESTED** [review]

File: dataikuapi/dssclient.py [hunk 2]
```diff
@@ -644,6 +644,95 @@ def create_global_api_key(self, label=None, description=None, admin=False):
...
+        """
+        Get a handle to interact with a specific Personal API key
+
+        :param str key: the secret key of the desired API key
+        :returns: A :class:`dataikuapi.dss.admin.DSSPersonalApiKey` API key handle
+        """
+        return DSSPersonalApiKey(self, key)
+
+
+    def create_personal_api_key(self, label="", description="", as_type='dict'):
```
[the remark this answers is not in the document]
Comment (qcastel) [reply dataikuapi/dssclient.py:674]: That is true. Should we authorise USER_FROM_UI as well then?

File: dataikuapi/dssclient.py [hunk 3]
```diff
@@ -644,6 +644,95 @@ def create_global_api_key(self, label=None, description=None, admin=False):
...
+        :returns: All personal API keys associated with your user
+        """
+        resp = self._perform_json(
+            "GET", "/personal-api-keys/")
+        if as_type == 'objects':
+            return [DSSPersonalApiKey(self, item['id']) for item in resp]
+        else:
+            return resp
+
+    def get_personal_api_key(self, key):
```
[the remark this answers is not in the document]
Comment (mhoarau) [reply dataikuapi/dssclient.py]: The doc now says: "the `secret` id".

**Review by FChataigner — APPROVED** [review]

**Review by mhoarau — CHANGES_REQUESTED** [review]

File: dataikuapi/dssclient.py [hunk 4]
```diff
@@ -644,6 +646,105 @@ def create_global_api_key(self, label=None, description=None, admin=False):
         key = resp.get('key', '')
         return DSSGlobalApiKey(self, key)
 
+    ########################################################
+    # Personal API Keys
+    ########################################################
+
+    def list_personal_api_keys(self, as_type='dict'):
```
[the remark this answers is not in the document]
Comment (qcastel) [reply dataikuapi/dssclient.py:675]: as discussed on slack, we keep dict as create_personal_api_key already existed and was returning dict by default.
For retro-compatiblity reason, we will not change the default return type

[the remark this answers is not in the document]
Comment (qcastel) [reply dataikuapi/dssclient.py]: actually after discussion with Fred and Mickael, it's been agreed that the new endpoints would have the default type 'object' but the create_personal_api_key will be the only one having `dict` as default for retro-compatiblity reason

**Review by FChataigner — APPROVED** [review]

File: dataikuapi/dssclient.py [hunk 5]
```diff
@@ -667,6 +670,105 @@ def create_global_api_key(self, label=None, description=None, admin=False):
         key = resp.get('key', '')
         return DSSGlobalApiKey(self, key)
 
+    ########################################################
+    # Personal API Keys
+    ########################################################
+
+    def list_personal_api_keys(self, as_type='object'):
```
Comment (mhoarau) [inline dataikuapi/dssclient.py:677]: Careful : mismatch between `object` and `objects` (also in the doc).
So I did not test but if we follow the doc and/or use the default value you'll get a `dict`.

**Review by mhoarau — CHANGES_REQUESTED** [review]

File: dataikuapi/dssclient.py (hunk 5 above)
Comment (qcastel) [reply dataikuapi/dssclient.py]: Good catch!

File: dataikuapi/dssclient.py [hunk 6]
```diff
@@ -667,6 +670,105 @@ def create_global_api_key(self, label=None, description=None, admin=False):
...
+        resp = self._perform_json(
+            "POST", "/personal-api-keys/", body={"label": label, "description": description})
+        if resp is None:
+            raise Exception('API key creation returned no data')
+        if resp.get('messages', {}).get('error', False):
+            raise Exception('API key creation failed : %s' % (json.dumps(resp.get('messages', {}).get('messages', {}))))
+        if not resp.get('id', False):
+            raise Exception('API key creation returned no key')
+
+        if as_type == 'objects':
```
Comment (mhoarau) [inline dataikuapi/dssclient.py:723]: IMO it should be plural only when a list of object is returned.

**Review by mhoarau — CHANGES_REQUESTED** [review]

File: dataikuapi/dssclient.py [hunk 7]
```diff
@@ -667,6 +670,105 @@ def create_global_api_key(self, label=None, description=None, admin=False):
...
 
+    ########################################################
+    # Personal API Keys
+    ########################################################
+
+    def list_personal_api_keys(self, as_type='objects'):
+        """
+        List all your personal API keys
+
+        :param str as_type: How to return the personal API keys. Possible values are "dict" and "object"
```
Comment (mhoarau) [inline dataikuapi/dssclient.py:681]: Still missing plural to object here (and other places)

**Review by mhoarau — CHANGES_REQUESTED** [review]

**Review by mhoarau — APPROVED** [review]

File: dataikuapi/dss/admin.py [hunk 8]
```diff
@@ -1169,6 +1171,38 @@ def delete(self):
         return self.client._perform_empty(
             "DELETE", "/personal-api-keys/%s" % self.id_)
 
+class DSSPersonalApiKeyListItem(DSSTaggableObjectListItem):
+    """An item in a list of datasets. Do not instantiate this class, :meth:`dataikuapi.DSSClient.list_personal_api_keys` or :meth:`dataikuapi.DSSClient.list_all_personal_api_keys`"""
```
Comment (mhoarau) [inline dataikuapi/dss/admin.py:1175]: String should be on multiple lines and better formatted.

Comment (mhoarau) [inline dataikuapi/dss/admin.py:1175]: it says "list of datasets".

File: dataikuapi/dss/admin.py [hunk 9]
```diff
@@ -1169,6 +1171,38 @@ def delete(self):
         return self.client._perform_empty(
             "DELETE", "/personal-api-keys/%s" % self.id_)
 
+class DSSPersonalApiKeyListItem(DSSTaggableObjectListItem):
```
Comment (mhoarau) [inline dataikuapi/dss/admin.py:1174]: An API key is not a `taggable` object. It does not have tags. 
FYI: `com.dataiku.dip.server.services.TaggableObjectsService.TaggableObject`
IMO: could simply extends a `dict`.
Should be

File: dataikuapi/dss/admin.py [hunk 10]
```diff
@@ -1136,6 +1138,71 @@ def set_definition(self, definition):
             body = definition)
 
 
+class DSSPersonalApiKey(object):
+    """
+    A personal API key on the DSS instance
+    Do not call that directly, use :meth:`dataikuapi.DSSClient.get_personal_api_key`
```
Comment (mhoarau) [inline dataikuapi/dss/admin.py:1144]: We need a coma between the two sentences.

File: dataikuapi/dss/admin.py (hunk 8 above)
Comment (mhoarau) [inline dataikuapi/dss/admin.py:1175]: We are missing a few words.
BTW we have in the 3 classes close to this one 3 different wording to say the same thing. We could at least chose one here and use it for the 3 of them:
> Do not create this object directly, use :meth:`my_method` instead.

WDYT?

File: dataikuapi/dssclient.py [hunk 11]
```diff
@@ -674,21 +674,24 @@ def create_global_api_key(self, label=None, description=None, admin=False):
     # Personal API Keys
     ########################################################
 
-    def list_personal_api_keys(self, as_type='objects'):
+    def list_personal_api_keys(self, as_type='listitems'):
         """
         List all your personal API keys.
 
         :param str as_type: How to return the personal API keys. Possible values are "dict" and "objects"
```
Comment (mhoarau) [inline dataikuapi/dssclient.py:681]: Still says `dict`. (same for the `all` method)

File: dataikuapi/dssclient.py [hunk 12]
```diff
@@ -674,21 +674,24 @@ def create_global_api_key(self, label=None, description=None, admin=False):
...
         
-        :return: if as_type=dict, each personal API keys is returned as a dict.
+        :return: if as_type=listitems, each one as a :class:`dataikuapi.dss.admin.DSSPersonalApiKeyListItem`.
```
Comment (mhoarau) [inline dataikuapi/dssclient.py:683]: Like the line below I would replace `one` by `key`. (same for the `all` method)

File: dataikuapi/dss/admin.py [hunk 13]
```diff
@@ -1136,6 +1138,71 @@ def set_definition(self, definition):
...
+        self.client = client
+
+    def to_personal_api_key(self):
+        """Gets the :class:`DSSPersonalApiKey` corresponding to this item"""
+        return DSSPersonalApiKey(self.client, self._data["id"])
+
+    @property
+    def id(self):
+        return self._data["id"]
+    @property
```
Comment (mhoarau) [inline dataikuapi/dss/admin.py:1187]: Missing blank line.

File: dataikuapi/dss/admin.py [hunk 14]
```diff
@@ -1136,6 +1138,71 @@ def set_definition(self, definition):
...
+        return self._data["key"]
+    @property
+    def label(self):
+        return self._data["label"]
+    @property
+    def description(self):
+        return self._data["description"]
+    @property
+    def created_on(self):
+        return self._data["createdOn"]
```
Comment (mhoarau) [inline dataikuapi/dss/admin.py:1201]: We tend to return a datetime more and more.
E.g. `dataikuapi.dss.admin.DSSUserActivity.last_successful_login`.

**Review by mhoarau — CHANGES_REQUESTED** [review]

File: dataikuapi/dss/admin.py (hunk 8 above)
Comment (qcastel) [reply dataikuapi/dss/admin.py:1175]: Good idea. I just aligned all the classes to it.

File: dataikuapi/dss/admin.py [hunk 15]
```diff
@@ -1136,12 +1137,89 @@ def set_definition(self, definition):
...
+        super(DSSPersonalApiKeyListItem, self).__init__(data)
+        self.client = client
+
+    def to_personal_api_key(self):
+        """Gets the :class:`DSSPersonalApiKey` corresponding to this item"""
+        return DSSPersonalApiKey(self.client, self._data["id"])
+
+    @property
+    def id(self):
+        return self._data["id"]
```
Comment (mhoarau) [inline dataikuapi/dss/admin.py]: Testing any properties leads to:
```py
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-825eb33a416e> in <module>()
      1 personal_keys = client.list_personal_api_keys(as_type="listitems")
      2 test = personal_keys[3]
----> 3 date = test.created_on
      4 date.isoformat()

/Users/mhoarau/devenv100/dataiku-api-client-python/dataikuapi/dss/admin.py in created_on(self)
   1207     @property
   1208     def created_on(self):
-> 1209         timestamp = self._data["createdOn"]
   1210         return datetime.datetime.fromtimestamp(timestamp / 1000) if timestamp > 0 else None
   1211 

AttributeError: 'DSSPersonalApiKeyListItem' object has no attribute '_data'
```

**Review by mhoarau — CHANGES_REQUESTED** [review]
List item is broken (and fixed by the previous commit). Can you double check there are no other issues?

File: dataikuapi/dss/admin.py (hunk 15 above)
Comment (qcastel) [reply dataikuapi/dss/admin.py]: Fixed by the commit pushed by Mickael directly on the PR

**Pull request merged**